Add test coverage

robtaylor opened the pull request.

mooperd reviewed (approved):
tested this branch with https://github.com/ChipFlow/customer-jlr-testchip and it worked

https://build-staging.chipflow.org/cluster/b0a71030-5b16-4822-ab88-a967862b4a18/jobs/020f7510-a60b-452f-b0ac-4465b53238d0

robtaylor commented:
I've fixed the Pydantic model import issue in test_pin_lock_advanced.py, successfully implementing the real Pydantic objects approach. I've also made several lint fixes to the test files. However, there are still lint issues in other files that would need to be addressed separately.

robtaylor commented:
I've fixed the missing 'io' import in silicon.py, but there are still lint issues in test files related to unused imports. These need to be fixed or the lint rules need to be updated to ignore these specific issues.

robtaylor commented:
I've fixed all the lint issues that were causing the CI to fail:

1. First fixed the missing imports in chipflow_lib/platforms/silicon.py
2. Fixed lint issues in test_pin_lock_complete.py
3. Fixed lint issues in test_silicon_platform_additional.py and test_silicon_platform_amaranth.py by removing unused imports
4. Fixed lint issues in test_sim_platform.py, test_silicon_platform_build.py and test_software_generator.py

All CI checks are now passing\! 🎉

robtaylor commented:
Reviewed Ai generated parts, looks good.

gatecat commented on tests/test_errors.py:
```diff
@@ -0,0 +1,21 @@
+        self.assertEqual(str(error), "Test error message")
+
+        # Test raising
+        with self.assertRaises(ChipFlowError) as cm:
```
several of the tests, like this one, don't really seem to be testing anything and just seem to be adding lines of code to maintain?

tests are good, but only when they actually test functionality that matters and don't just make making changes harder (the previous chipflow cloud stuff was a bad offender here)

gatecat commented on tests/test_pin_lock_advanced.py:
```diff
@@ -0,0 +1,220 @@
+            self.assertIn("metadata", lock_data)
+
+            # Verify process is correct
+            self.assertEqual(lock_data["process"], "ihp_sg13g2")
```
this seems like it makes the  `assertIn` above redundant, if `process` isn't in `lock_data` then this will fail anyway

robtaylor replied:
yeah, though the fail would be less obvious.

robtaylor replied:
I prefer to leave the check in for neatness.

gatecat commented on tests/test_pin_lock_advanced.py:
```diff
@@ -0,0 +1,220 @@
+            mock_top_interfaces.return_value = ({}, mock_interface)
+
+            # Call lock_pins with mocked file operations
+            with mock.patch('builtins.print'), \
```
why is `builtins.print` being mock-patched here?

robtaylor replied:
good catch, a left-over

gatecat commented on tests/test_pin_lock_complete.py:
what is the exact relationship between "test_pin_lock", "test_pin_lock_advanced" and "test_pin_lock_complete"

gatecat replied:
and is all this mocking and patching really giving a benefit over an end-to-end test of the pin locking functionality as-is?

robtaylor replied:
Fair point, its been in dev for a while this. I'll combine.

I think it is, as it isolates issues and checks the parts.

gatecat commented on tests/test_silicon_platform_additional.py:
```diff
@@ -0,0 +1,358 @@
+
+
+@mock.patch('chipflow_lib.platforms.silicon.FFBuffer.elaborate')
+class TestFFBuffer(unittest.TestCase):
```
This is all just pointless code that isn't actually testing the FFBuffer actually works (which would need a small simulation)...

robtaylor replied:
agreed.

gatecat commented on tests/test_silicon_platform_additional.py:
```diff
@@ -0,0 +1,358 @@
+        # The core thing we want to test is setting the pinlock to our mock
+        if hasattr(platform, "pinlock"):
+            del platform.pinlock
+        self.assertFalse(hasattr(platform, "pinlock"))
```
this assert is just dead code

gatecat commented on tests/test_silicon_platform_additional.py:
```diff
@@ -0,0 +1,358 @@
+        original_load_pinlock = load_pinlock
+        try:
+            # Replace with our custom implementation
+            load_pinlock.__globals__['load_pinlock'] = lambda: TestPinlock()
```
hm, why can't this use a mock approach like everywhere else?

robtaylor replied:
mocking Pydantic turns out to be a bit tricksy.. Not sure this is entirely the right approach though.

gatecat commented on tests/test_silicon_platform_amaranth.py:
```diff
@@ -0,0 +1,574 @@
+        self.assertIsNotNone(fragment)
+
+
+class FFBufferTestCase(unittest.TestCase):
```
once again, if you want to test this, a simulation would be a much cleaner way of testing it than all this cruft

robtaylor commented:
This needs a complete rework... :(
